### PR TITLE
Improve `keyword` parsing

### PR DESCRIPTION
**Pull request by shantanu-sardesai**
Resolves https://github.com/jank-lang/jank/issues/204, https://github.com/jank-lang/jank/issues/578.

Now jank also handles the auto-resolved qualified keywords properly:

```jank
user=> '(::)
─ lex/invalid-keyword ──────────────────────────────────────────────────────────────────────────────
error: An auto-resolved keyword may not start with ')'.

─────┬──────────────────────────────────────────────────────────────────────────────────────────────
     │ /var/folders/xd/9v1dcg0d46l7d19830vq0lhc0000gn/T/jank-repl-Q7azuw
─────┼──────────────────────────────────────────────────────────────────────────────────────────────
  1  │ '(::)
     │   ^ Found here.
─────┴──────────────────────────────────────────────────────────────────────────────────────────────
```

### CODE REVIEW

File: compiler+runtime/src/cpp/jank/read/lex.cpp [hunk 1]
```diff
@@ -1205,20 +1205,33 @@ namespace jank::read::lex
           {
             ++pos;
             auto const after_dd_colon(peek());
-            if(after_dd_colon.is_ok() && after_dd_colon.expect_ok().character == '/')
+
+            /* Invalid ":: " source found. */
+            if(after_dd_colon.is_err())
+            {
+              return error::lex_invalid_keyword(
+                "A auto-resolved keyword must contain a valid symbol after the '::'.",
```
Comment (jeaye) [inline compiler+runtime/src/cpp/jank/read/lex.cpp]: `An`, not `A` here, since `auto` starts with a vowel sound.

File: compiler+runtime/src/cpp/jank/read/lex.cpp [hunk 2]
```diff
@@ -1205,20 +1205,33 @@ namespace jank::read::lex
...
+                if(result.is_err() || !is_symbol_char(ch))
                 {
                   break;
                 }
                 pos += result.expect_ok().len;
               }
-              return error::lex_invalid_keyword("An auto-resolved keyword may not start with '/'.",
-                                                { token_start, pos });
+              return error::lex_invalid_keyword(
+                util::format("An auto-resolved keyword may not start with '{}'.", ch),
```
Comment (jeaye) [inline compiler+runtime/src/cpp/jank/read/lex.cpp]: We have the information here for a better error. We know that we have a `::` and that we have one of two cases:

1. `::/`, so we can provide an error saying the namespace cannot be empty for the keyword.
2. `::` followed by a non-symbol char, so we can provide an error saying there's an unexpected end to that keyword. This is an incomplete keyword.

I think trying to word it as what a keyword can't start with is less helpful to the user. The user isn't going to type `::)` because they think a keyword can start with that. They're going to do that because they forgot to provide the rest of the keyword.

**Review by jeaye — CHANGES_REQUESTED** [review]
Looks good, just some error message improvements we can do.

File: compiler+runtime/src/cpp/jank/read/lex.cpp (hunk 2 above)
Comment (shantanu-sardesai) [reply compiler+runtime/src/cpp/jank/read/lex.cpp]: > ::/, so we can provide an error saying the namespace cannot be empty for the keyword.

Just wondering, since this is an auto-resolved keyword, can they even provide a namespace? If they can't then a better error would be to just say that you can't provide one, no?

Comment (jeaye) [reply compiler+runtime/src/cpp/jank/read/lex.cpp]: Yes they can. Auto-resolved keyword namespaces match namespace aliases (that sentence is a tongue twister).

For example:
```clojure
(require '[jank.compiler :as c])
::c/foo ; :jank.compiler/foo
```

File: compiler+runtime/test/cpp/jank/read/lex.cpp [hunk 3]
```diff
@@ -1810,6 +1811,33 @@ namespace jank::read::lex
...
+                make_error(kind::lex_invalid_keyword, 0, 2),
+              }));
+      }
+
+      SUBCASE("Comma after ::")
+      {
+        processor p{ "::," };
+        native_vector<jtl::result<token, error_ref>> const tokens(p.begin(), p.end());
+        CHECK(tokens == make_results({ make_error(kind::lex_invalid_keyword, 0, 2) }));
+      }
```
Comment (jeaye) [inline compiler+runtime/test/cpp/jank/read/lex.cpp:1850]: The lexer needs to be able to correctly and gracefully continue lexing on a bad token and we need to test that for each bad token. So please include a test like `:: 12` to ensure you get both an invalid keyword and a valid number with the value 12.

Comment (shantanu-sardesai) [reply compiler+runtime/test/cpp/jank/read/lex.cpp:1850]: Done.

File: compiler+runtime/test/cpp/jank/read/lex.cpp [hunk 4]
```diff
@@ -1810,6 +1811,44 @@ namespace jank::read::lex
...
+
+      SUBCASE("Graceful lexing after invalid ::")
+      {
+        processor p{ ":: 42" };
+        native_vector<jtl::result<token, error_ref>> const tokens(p.begin(), p.end());
+        CHECK(tokens
+              == make_results({
+                make_error(kind::lex_invalid_keyword, 0, 2),
+                token{ 0, 3, token_kind::integer, "42" }
```
Comment (jeaye) [inline compiler+runtime/test/cpp/jank/read/lex.cpp]: If this test passes, we have a lexer bug. This says that the integer token starts at offset 0 and is 3 bytes long. **These numbers matter.** When we fill in a test, we have to check them and make sure they make sense.

These do not make sense.

Also, the value for the integer should not be a string. It should be an integral value. My guess is you didn't run this locally and it'll fail in CI.

Comment (shantanu-sardesai) [reply compiler+runtime/test/cpp/jank/read/lex.cpp]: Ah, I thought the 0 represented the line number.

Weird, I did run it locally before committing, I'll check.

Comment (jeaye) [reply compiler+runtime/test/cpp/jank/read/lex.cpp]: I believe you, since CI also passed. :grin:  So that means we have a lexer bug with this change. We need to make sure to properly progress the position when returning an error.

Comment (shantanu-sardesai) [reply compiler+runtime/test/cpp/jank/read/lex.cpp]: This is weirder that I thought, I just updated the test case to:

```c++
token{ 3, 2, token_kind::integer, 42ll }
```

And it worked... Looking into it.

Comment (shantanu-sardesai) [reply compiler+runtime/test/cpp/jank/read/lex.cpp]: As I would expect, this fails:

```C++
CHECK(tokens.at(1).expect_ok() == token{ 0, 2, token_kind::integer, 42ll });
```

and this passes:

```C++
CHECK(tokens.at(1).expect_ok() == token{ 3, 2, token_kind::integer, 42ll });
```

> Corollary of the above point, no lexer bug 🙂.

Are we sure that the use of `==` to compare `tokens` (type: `native_vector<jtl::result<token, error_ref>>`) and output of `make_results` (type: `std::array<jtl::result<token, error_ref>, N>`) correct? Or something else is going weird/wrong or just me thinking incorrectly...

I saw it being used [here](https://github.com/shantanu-sardesai/jank/blob/2d822586734ed6b1a2e0045065c34e3317cf96c4/compiler+runtime/test/cpp/jank/read/lex.cpp#L666-L674) so I wrote my tests similarly.

Comment (jeaye) [reply compiler+runtime/test/cpp/jank/read/lex.cpp]: Yeah, looks like we have some bad `==` logic going on here! Try jumping to definition on the `==` usage here and working through the logic. You can also step through it with a debugger.

From what I can see, your usage is correct. Could be we've had an issue in there already and we're only spotting it now.

Comment (shantanu-sardesai) [reply compiler+runtime/test/cpp/jank/read/lex.cpp]: [Fixed the bug](https://github.com/jank-lang/jank/pull/577/commits/476b9e78aee163908a957073f974546635befb37). As I feared, it opened up (or maybe exposed) a can of worms...

```
===============================================================================
/Users/shantanusardesai/Desktop/code/projects/jank/compiler+runtime/test/cpp/jank/read/lex.cpp:492:
TEST SUITE: lex
TEST CASE:  Ratio
  Failures - x.x/x

===============================================================================
/Users/shantanusardesai/Desktop/code/projects/jank/compiler+runtime/test/cpp/jank/read/lex.cpp:492:
TEST SUITE: lex
TEST CASE:  Ratio
  Failures - xex/x

===============================================================================
/Users/shantanusardesai/Desktop/code/projects/jank/compiler+runtime/test/cpp/jank/read/lex.cpp:601:
TEST SUITE: lex
TEST CASE:  Integer
  Invalid octal number

===============================================================================
/Users/shantanusardesai/Desktop/code/projects/jank/compiler+runtime/test/cpp/jank/read/lex.cpp:601:
TEST SUITE: lex
TEST CASE:  Integer
  Edge case: Invalid mixed radix

===============================================================================
/Users/shantanusardesai/Desktop/code/projects/jank/compiler+runtime/test/cpp/jank/read/lex.cpp:601:
TEST SUITE: lex
TEST CASE:  Integer
  Mixed valid and invalid numbers

===============================================================================
/Users/shantanusardesai/Desktop/code/projects/jank/compiler+runtime/test/cpp/jank/read/lex.cpp:601:
TEST SUITE: lex
TEST CASE:  Integer
  Invalid hex numbers

===============================================================================
/Users/shantanusardesai/Desktop/code/projects/jank/compiler+runtime/test/cpp/jank/read/lex.cpp:601:
TEST SUITE: lex
TEST CASE:  Integer
  Invalid hex - 2.0x1

===============================================================================
/Users/shantanusardesai/Desktop/code/projects/jank/compiler+runtime/test/cpp/jank/read/lex.cpp:601:
TEST SUITE: lex
TEST CASE:  Integer
  Invalid arbitrary radix

===============================================================================
/Users/shantanusardesai/Desktop/code/projects/jank/compiler+runtime/test/cpp/jank/read/lex.cpp:980:
TEST SUITE: lex
TEST CASE:  N suffix
  1N/34

===============================================================================
/Users/shantanusardesai/Desktop/code/projects/jank/compiler+runtime/test/cpp/jank/read/lex.cpp:1108:
TEST SUITE: lex
TEST CASE:  Real
  Scientific notation
  Missing exponent

===============================================================================
/Users/shantanusardesai/Desktop/code/projects/jank/compiler+runtime/test/cpp/jank/read/lex.cpp:1108:
TEST SUITE: lex
TEST CASE:  Real
  Scientific notation
  Signs after exponent found

===============================================================================
/Users/shantanusardesai/Desktop/code/projects/jank/compiler+runtime/test/cpp/jank/read/lex.cpp:1108:
TEST SUITE: lex
TEST CASE:  Real
  Scientific notation
  Extra dots

===============================================================================
/Users/shantanusardesai/Desktop/code/projects/jank/compiler+runtime/test/cpp/jank/read/lex.cpp:1108:
TEST SUITE: lex
TEST CASE:  Real
  Scientific notation
  Extra characters in exponent

===============================================================================
/Users/shantanusardesai/Desktop/code/projects/jank/compiler+runtime/test/cpp/jank/read/lex.cpp:1297:
TEST SUITE: lex
TEST CASE:  Big decimal
  Invalid - M in middle

===============================================================================
[doctest] test cases:  114 |  109 passed |  5 failed | 0 skipped
[doctest] assertions: 1665 | 1651 passed | 14 failed |
[doctest] Status: FAILURE!
```

Shall I fix these in the PR as well? I'm assuming yes, since we don't want to merge without a green pipeline?

Comment (shantanu-sardesai) [reply compiler+runtime/test/cpp/jank/read/lex.cpp]: > Ah, I thought the 0 represented the line number.

Heh, seems like there were similar misunderstandings in the [tests](https://github.com/jank-lang/jank/pull/577/commits/b5cb3eea633b372503ff48c32b6d751a96bb4585) on what the numbers mean 🙂. Anyways, fixed all of them.

Comment (jeaye) [reply compiler+runtime/test/cpp/jank/read/lex.cpp]: Nice! Yeah, we want all of these to be fixed. Thanks for putting in that effort.

File: compiler+runtime/test/cpp/jank/read/lex.cpp [hunk 5]
```diff
@@ -78,7 +78,12 @@ namespace jank::read::lex
 
         auto const &lhs_err(lhs.expect_err());
         auto const &rhs_err(rhs.expect_err());
-        return lhs_err->kind == rhs_err->kind && lhs_err->source == rhs_err->source;
+        auto const is_eq{ lhs_err->kind == rhs_err->kind && lhs_err->source == rhs_err->source };
+
+        if(!is_eq)
+        {
+          return false;
+        }
```
Comment (jeaye) [inline compiler+runtime/test/cpp/jank/read/lex.cpp:86]: From what I see we don't actually return a value if `is_eq`. Shouldn't there be a `return true;` at the end of the fn, if you want to do this?

**Review by jeaye — CHANGES_REQUESTED** [review]

File: compiler+runtime/test/cpp/jank/read/lex.cpp (hunk 5 above)
Comment (shantanu-sardesai) [reply compiler+runtime/test/cpp/jank/read/lex.cpp:86]: [We already have that](https://github.com/jank-lang/jank/blob/main/compiler+runtime/test/cpp/jank/read/lex.cpp#L96), no?

If `is_eq` then we wish to check exhaustively and at the end of the loop, since we never bailed out in between, we just return `true`.

Comment (jeaye) [reply compiler+runtime/test/cpp/jank/read/lex.cpp:86]: Ah, yes. It was hidden.

image

**Comment by jeaye** [comment]
Thank you, Shantanu!
